### ISSUE TITLE
Query pending operations (on nic)

### ISSUE THREAD

**Issue by zenhack** [CCI-MOC/hil#486]
We should be able to query the existence of pending network operations on a
switch. This would be useful for users in general, but is also something that is
important for implementing recursive HaaS correctly.

@gsilvis, @SahilTikale, @okrieg, @lokI8 and I just had a conversation about the
async questions that needed to be resolved, and what we settled on was this:
We'd add this query api call, and then the recursive HaaS switch driver would
invoke the base HaaS's api to do the network operating, and use this call to
poll, not returning until the operating has gone through.

@gsilvis has some misgivings about fudging it to look synchronous (I'll let him
describe them in more detail), but agrees that we can go ahead with this for
now

**Comment by gsilvis** [comment]
Here's my thoughts on it:

The way we call into the driver right now, we call it with a single network operation [aka a port and a channel configuration], and it completes everything in that call.  This works okay for recursive HaaS:  in that function call, it will make the API call on the base HaaS, and then poll the network operation until it completes.

However, this essentially synchronous-izes the base HaaS's API: the recursive HaaS can only do one of these calls at once.  I was quoted 10 seconds for the duration of a synchronous HaaS API call.  This means the recursive HaaS can only service one API call per 10 seconds.

Firstly, this is clearly too slow a rate to be processing API calls at.  So that needs to be optimized.  I think we can get it down to much shorter than that.  Secondly, I would argue that, if we're essentially forced to block until an operation completes anyways, we might as well make the operation synchronous.  [On the other hand, there's a big gotcha here---if the HaaS crashes, then it needs to redo all of the operations that it might or might not have completed.  This will involve doing operations whose API call did not return to the user, which is a little scary.]


Here's an option that actually makes use of the asynchronousness of the HaaS API:

We need three services running.  The API server takes in API calls, and puts them in the operation queue.  The network application service looks at things in the operation queue, and does them: in the case of recursive HaaS, it makes the initial API call to the base HaaS, and puts the operation in a list of operations to keep checking the status of.  The third service, the 'transaction status checking' service, then polls all of its transactions until they are complete.  When a transaction completes, it then (finally) updates the database state.

This still guarantees that operations are processed serially: if API call A reaches the recursive HaaS before API call B, then the network application service will make API call A to base HaaS before it makes API call B.  Then base HaaS is guaranteed to service API call A before API call B as well.

**Comment by zenhack** [comment]
Implementation issue I just noticed: both the current model and the one you propose assume that applying an operation is idempotent; it's therefore safe to reapply if we crash in the middle of a db transaction. for this to work the case of rHaaS, for this to work you'd have to respond to an error from the base haas by waitning for the queue to empty, and then submitting again.

But yeah, I think what you propose makes sense in the long term (might be nice to have two threads in one process, rather than have *another* daemon to start, but the basic idea is fine). I think we should go ahead with the current solution for now.

**Comment by henn** [comment]
I collected some thoughts into [an async proposal](https://github.com/CCI-MOC/haas/wiki/Asynchronous-operations). Please comment everyone!

**Comment by naved001** [comment]
Working on this [here](https://github.com/naved001/hil/tree/query-networking-actions)

**Comment by naved001** [comment]
The basic stuff is there now, still needs some more work and tests. Here's [the diff](https://github.com/CCI-MOC/hil/compare/master...naved001:query-networking-actions?expand=1).

Some questions:
1. Once an action is complete, we mark it's status as DONE or ERROR. The action is obviously not deleted from the table. We need to decide for how long are we going to keep that action in the table. @apoorvemohan suggested that once a user checks for the status of their operation and if it was DONE or ERROR, we then delete the entry from the table since it's safe to assume that the user has the information they need.
We could delete all entries, with status==DONE|ERROR after some configurable time, whether or not their status was checked. 

2. Right now, we are using the same networking action table to store completed Actions. So when queuing a new action, we have to first filter by nic, and see whether there's any pending action on that nic (because there could be DONE or ERROR). This could potentially make things slow (again, pointed out by @apoorvemohan). One way to avoid this is to move completed actions to a new table and keep the OG networking action table clean. 

Depending on what we decide for 1 and 2, I'll make the changes and then add appropriate migration scripts, new tests, fix old tests etc.

cc: @zenhack

**Comment by zenhack** [comment]
Two more ideas re: (1):

1. Just delete DONE actions, don't keep them at all. Users should interpret missing actions as completed. Successful responses from the api calls that create actions indicate that the action has been inserted durably into the queue, so this is always valid. This is in keeping with keeping HIL simple; remember, it's supposed to be a low-level tool. If we want something more user friendly that knows about actions that are "done", we can build that outside of the (priviledged) HIL process.
2. Delete DONE actions when a new action is added on the same nic. This keeps the number of database objects bounded to a managable number.

I like (1) better personally.

Either strategy for (1) should keep (2) from being a concern, since they will prevent us from having "too many" actions to look through.

**Comment by naved001** [comment]
If we delete DONE actions, and then a user calls `get_status` on the UUID for that action that would return a 404 because it's not found, and inferring from that the it's complete is a little weird. 

(2) sounds more reasonable to me. Just delete the previous action if status was DONE or ERROR, when inserting a new one. It takes care of having "too many" actions, and only keeps the status of the most recent operation on that nic which probably is the relevant operation.
This won't keep history, but like you mentioned, hil is supposed to be a low-level tool, we don't need to maintain a history.

If we agree on this, I would go ahead and do (2).  Thanks for the suggestions!

**Comment by zenhack** [comment]
If you prefer (2) I'm fine with that.
